src/L1/rollup/ScrollChain.sol

```diff
@@ -575,12 +649,151 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
      * Internal Functions *
      **********************/
 
+    /// @dev Internal function to do common checks before actual batch committing.
+    /// @param _parentBatchHeader The parent batch header in calldata.
+    /// @param _chunks The list of chunks in memory.
+    /// @return _parentBatchHash The batch hash of parent batch header.
+    /// @return _batchIndex The index of current batch.
+    /// @return _totalL1MessagesPoppedOverall The total number of L1 messages popped before current batch.
+    function _beforeCommitBatch(bytes calldata _parentBatchHeader, bytes[] memory _chunks)
+        private
+        view
+        returns (
+            bytes32 _parentBatchHash,
+            uint256 _batchIndex,
+            uint256 _totalL1MessagesPoppedOverall
+        )
+    {
+        // check whether the batch is empty
+        if (_chunks.length == 0) revert ErrorBatchIsEmpty();
+        (, _parentBatchHash, _batchIndex, _totalL1MessagesPoppedOverall) = _loadBatchHeader(_parentBatchHeader);
+        unchecked {
+            _batchIndex += 1;
+        }
+        if (committedBatches[_batchIndex] != 0) revert ErrorBatchIsAlreadyCommitted();
+    }
+
+    /// @dev Internal function to do common checks after actual batch committing.
+    /// @param _batchIndex The index of current batch.
+    /// @param _batchHash The hash of current batch.
+    function _afterCommitBatch(uint256 _batchIndex, bytes32 _batchHash) private {
+        committedBatches[_batchIndex] = _batchHash;
+        emit CommitBatch(_batchIndex, _batchHash);
+    }
+
+    /// @dev Internal function to do common checks before actual batch finalization.
+    /// @param _batchHeader The current batch header in calldata.
+    /// @param _postStateRoot The state root after current batch.
+    /// @return batchPtr The start memory offset of current batch in memory.
+    /// @return _batchHash The hash of current batch.
+    /// @return _batchIndex The index of current batch.
+    function _beforeFinalizeBatch(bytes calldata _batchHeader, bytes32 _postStateRoot)
+        internal
+        view
+        returns (
+            uint256 batchPtr,
+            bytes32 _batchHash,
+            uint256 _batchIndex
+        )
+    {
+        if (_postStateRoot == bytes32(0)) revert ErrorStateRootIsZero();
+
+        // compute batch hash and verify
+        (batchPtr, _batchHash, _batchIndex, ) = _loadBatchHeader(_batchHeader);
```

Comment:
```suggestion
        (_, _, _batchIndex, ) = _loadBatchHeader(_batchHeader);
```

Reply:
these two are returned and will be used by the caller.

Reply:
got it.